akka-typed/src/main/scala/akka/typed/ScalaDSL.scala

```diff
@@ -0,0 +1,399 @@
+
+    @tailrec private def run(ctx: ActorContext[T], next: Behavior[T]): Behavior[T] = {
+      setBehavior(ctx, next)
+      if (inbox.hasMessages) run(ctx, behavior.message(ctx, inbox.receiveMsg()))
+      else if (isUnhandled(next)) Unhandled
+      else if (isAlive(next)) this
+      else Stopped
+    }
+
+    override def management(ctx: ActorContext[T], msg: Signal): Behavior[T] =
+      run(ctx, behavior.management(ctx, msg))
+    override def message(ctx: ActorContext[T], msg: T): Behavior[T] =
+      run(ctx, behavior.message(ctx, msg))
+
+    override def toString: String = s"SynchronousSelf($behavior)"
+  }
+
+  /**
+   * A behavior combinator that feeds incoming messages and signals both into
+   * the left and right sub-behavior and allows them to evolve independently of
+   * each other. When one of the sub-behaviors terminates the other takes over
+   * exclusively. When both sub-behaviors respond to a [[Failed]] signal, the
+   * response with the higher precedence is chosen (see [[Failed$]]).
+   */
+  case class And[T](left: Behavior[T], right: Behavior[T]) extends Behavior[T] {
+
+    override def management(ctx: ActorContext[T], msg: Signal): Behavior[T] = {
+      val l = left.management(ctx, msg)
+      val r = right.management(ctx, msg)
+      if (isUnhandled(l) && isUnhandled(r)) Unhandled
+      else {
+        val nextLeft = canonicalize(ctx, l, left)
+        val nextRight = canonicalize(ctx, r, right)
+        val leftAlive = isAlive(nextLeft)
+        val rightAlive = isAlive(nextRight)
+
+        if (leftAlive && rightAlive) And(nextLeft, nextRight)
+        else if (leftAlive) nextLeft
+        else if (rightAlive) nextRight
+        else Stopped
+      }
+    }
+
+    override def message(ctx: ActorContext[T], msg: T): Behavior[T] = {
+      val l = left.message(ctx, msg)
+      val r = right.message(ctx, msg)
+      if (isUnhandled(l) && isUnhandled(r)) Unhandled
+      else {
+        val nextLeft = canonicalize(ctx, l, left)
+        val nextRight = canonicalize(ctx, r, right)
+        val leftAlive = isAlive(nextLeft)
+        val rightAlive = isAlive(nextRight)
+
+        if (leftAlive && rightAlive) And(nextLeft, nextRight)
+        else if (leftAlive) nextLeft
+        else if (rightAlive) nextRight
+        else Stopped
+      }
+    }
+  }
+
+  /**
+   * A behavior combinator that feeds incoming messages and signals either into
+   * the left or right sub-behavior and allows them to evolve independently of
+   * each other. The message or signal is passed first into the left sub-behavior
+   * and only if that results in [[Behavior$.Unhandled]] is it passed to the right
+   * sub-behavior. When one of the sub-behaviors terminates the other takes over
+   * exclusively. When both sub-behaviors respond to a [[Failed]] signal, the
+   * response with the higher precedence is chosen (see [[Failed$]]).
+   */
+  case class Or[T](left: Behavior[T], right: Behavior[T]) extends Behavior[T] {
```

Comment:
I was thinking about adding an `Escape(nextBehavior)` return that either side could use to shed the `Or`, but a `Join` makes sense as well. This is precisely the reason why this is not yet included :-)